python_modules/libraries/dagster-powerbi/dagster_powerbi/resource.py

```diff
@@ -233,6 +234,68 @@ def build_defs(
         )
 
 
+def _workspace_data_to_cacheable_data(
+    workspace_data: PowerBIWorkspaceData,
+) -> Sequence[Mapping[str, Any]]:
+    return [
+        data.to_cached_data()
+        for data in [
+            *workspace_data.dashboards_by_id.values(),
+            *workspace_data.reports_by_id.values(),
+            *workspace_data.semantic_models_by_id.values(),
+            *workspace_data.data_sources_by_id.values(),
+        ]
+    ]
+
+
+def _build_assets_defs_from_workspace_data(
+    workspace_data: PowerBIWorkspaceData,
+    workspace: PowerBIWorkspace,
+    translator_cls: Type[DagsterPowerBITranslator],
+    enable_refresh_semantic_models: bool,
+) -> Sequence[AssetsDefinition]:
+    translator = translator_cls(context=workspace_data)
+
+    if enable_refresh_semantic_models:
+        all_external_data = [
+            *workspace_data.dashboards_by_id.values(),
+            *workspace_data.reports_by_id.values(),
+            *workspace_data.data_sources_by_id.values(),
+        ]
+        all_executable_data = [*workspace_data.semantic_models_by_id.values()]
+    else:
+        all_external_data = [
+            *workspace_data.dashboards_by_id.values(),
+            *workspace_data.reports_by_id.values(),
+            *workspace_data.data_sources_by_id.values(),
+            *workspace_data.semantic_models_by_id.values(),
+        ]
+        all_executable_data = []
+
+    all_external_asset_specs = [translator.get_asset_spec(content) for content in all_external_data]
+    all_executable_asset_specs = [
+        translator.get_asset_spec(content) for content in all_executable_data
+    ]
+
+    executable_assets = []
+    for content, spec in zip(all_executable_data, all_executable_asset_specs):
+        dataset_id = content.properties["id"]
+
+        @multi_asset(
+            specs=[spec],
+            name="_".join(spec.key.path),
+            resource_defs={"power_bi": workspace.get_resource_definition()},
+        )
+        def asset_fn(context: AssetExecutionContext, power_bi: PowerBIWorkspace) -> None:
+            power_bi.trigger_refresh(dataset_id)
+            power_bi.poll_refresh(dataset_id)
+            context.log.info("Refresh completed.")
+
+        executable_assets.append(asset_fn)
+
+    return [*external_assets_from_specs(all_external_asset_specs), *executable_assets]
+
+
 class PowerBICacheableAssetsDefinition(CacheableAssetsDefinition):
```

Comment:
can't we just kill this?

Reply:
as discussed offline not rming in this pr